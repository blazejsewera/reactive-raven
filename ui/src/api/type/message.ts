export type ApiMessage = {
  username: string
  timestamp: string
  title: string
  subtitle?: string
  body?: string
  id?: string
}

// example:
// { "username": "Test", "timestamp": "a", "title": "Test title", "subtitle": "Test subtitle", "body": "Test body.", "id": "1234" }
