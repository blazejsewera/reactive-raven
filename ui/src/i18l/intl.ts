/**
 * Intl is an interface for an internationalization (i18l) tool
 * that can provide translations for page.
 */
export interface Intl {
  getTranslation: (text: string) => string
}

export const intl: Intl = { getTranslation: text => text }
