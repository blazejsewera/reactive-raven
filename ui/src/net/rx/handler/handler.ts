import { Observable } from '../../../external/rxjs'
import { Action } from '../../../store/action/action'

export type Handler = (url: string) => Observable<Action>
