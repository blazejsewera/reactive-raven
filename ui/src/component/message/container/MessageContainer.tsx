import * as React from 'react'
import type { FC } from '../../../type/react'
import type { Message } from '../../../type/message'
import type { Intl } from '../../../i18l/intl'
import { cx } from '../../../util/cx'
import { Header } from './section/Header'
import { MessageCardList } from './section/MessageCardList'
import type { MessageContainerHandlers } from '../../../type/handler'

export interface MessageContainerProps {
  title: string
  messages: Message[]
  handlers: MessageContainerHandlers
  intl: Intl
  style?: string
}

export const MessageContainer: FC<MessageContainerProps> = ({ title, messages, handlers, intl, style }) => {
  const { onClear } = handlers
  return (
    <div
      className={cx(
        'bg-gray-100',
        'dark:bg-gray-700',
        'inline-block',
        'p-5',
        'pb-6',
        'rounded-3xl',
        'shadow-xl',
        style ?? '',
      )}
    >
      <Header title={title} onClear={onClear} />
      <MessageCardList {...{ messages, intl }} />
    </div>
  )
}
