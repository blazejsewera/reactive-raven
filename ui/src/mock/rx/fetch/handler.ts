import { observableOf } from '../../../external/rxjs'
import { Handler } from '../../../rx/handler/handler'
import { actionFetchSuccess } from '../../../store/action/action'
import { full, fullWithImage, partial, minimal, anotherUser } from '../../message.mock'

const messages = [full, fullWithImage, partial, minimal, anotherUser]

export const handler: Handler = () => observableOf(actionFetchSuccess(messages))
