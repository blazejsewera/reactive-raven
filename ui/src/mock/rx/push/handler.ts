import { of as observableOf, delay } from 'rxjs'
import type { PushHandler } from '../../../rx/push/handler'
import { actionPush } from '../../../store/action/action'
import type { Message } from '../../../type/message'

const message: Message = {
  username: 'Pushed',
  timestamp: '2',
  title: 'Pushed Message',
  id: 'pushed1',
}

const action = actionPush(message)

export const handler: PushHandler = () => observableOf(action).pipe(delay(1500))
