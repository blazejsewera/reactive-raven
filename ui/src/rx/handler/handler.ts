import type { Observable } from '../../external/rxjs'
import type { Action } from '../../store/action/action'

export type Handler = (url: string) => Observable<Action>
