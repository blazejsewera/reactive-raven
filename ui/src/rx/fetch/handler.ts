import { switchMap, catchError, map, observableFrom, observableOf } from '../../external/rxjs'
import type { Observable } from 'rxjs'
import type { Action } from '../../store/action/action'
import type { ApiMessage } from '../../api/type/message'
import type { Message } from '../../type/message'
import { actionFetchFail, actionFetchSuccess } from '../../store/action/action'
import { fromFetch as observableFromFetch } from 'rxjs/fetch'
import { updateTimeAll } from '../../util/message/time'

const SERVER_ERROR_MSG = 'Server error when fetching messages: '
const NETWORK_ERROR_MSG = 'Error when fetching messages: '

const toMessages = (payload: ApiMessage[]): Message[] =>
  payload.map(m => ({
    username: m.username,
    timestamp: m.timestamp,
    title: m.title,
    subtitle: m.subtitle,
    body: m.body,
    id: m.id,
  }))

const toAction = (payload: ApiMessage[]): Action => actionFetchSuccess(updateTimeAll(toMessages(payload)))
const observableActionSuccess = (response: Response): Observable<Action> =>
  observableFrom(response.json()).pipe(map(toAction))

const observableActionFail = (message: string): Observable<Action> => observableOf(actionFetchFail(message))

export type FetchHandler = (requestUrl: string) => Observable<Action>
export const handler: FetchHandler = requestUrl =>
  observableFromFetch(requestUrl).pipe(
    switchMap(
      (response): Observable<Action> =>
        response.ok ? observableActionSuccess(response) : observableActionFail(SERVER_ERROR_MSG + response.status),
    ),
    catchError(err => observableActionFail(NETWORK_ERROR_MSG + err?.message)),
  )
