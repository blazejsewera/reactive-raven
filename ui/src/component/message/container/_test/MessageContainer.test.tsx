import * as React from 'react'
import { create as render } from 'react-test-renderer'
import { Message } from '../../../../type/message'
import { intlMock } from '../../../../mock/intl.mock'
import { MessageContainer } from '../MessageContainer'
import { full, minimal } from '../../../../mock/message.mock'
import { MessageContainerHandlers } from '../../../../type/handler'

describe('MessageContainer component', () => {
  // given
  const intl = intlMock
  const handlers: MessageContainerHandlers = { onClear: jest.fn() }

  const testMessageContainer = (messages: Message[]) => {
    const tree = render(<MessageContainer {...{ title: 'TestTitle', messages, intl, handlers }} />).toJSON()
    expect(tree).toMatchSnapshot()
  }

  it('renders correctly with no messages', () => {
    testMessageContainer([])
  })
  it('renders correctly with one message', () => {
    testMessageContainer([full])
  })
  it('renders correctly with multiple messages', () => {
    testMessageContainer([full, minimal])
  })
})
