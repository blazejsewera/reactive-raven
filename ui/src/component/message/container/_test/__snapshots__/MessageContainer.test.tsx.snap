// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageContainer component renders correctly with multiple messages 1`] = `
<div
  className="bg-gray-100 dark:bg-gray-700 inline-block p-5 pb-6 rounded-3xl shadow-xl "
>
  <div
    className="flex mb-4 mx-2"
  >
    <p
      className="text-2xl font-bold text-gray-800 dark:text-gray-300"
    >
      TestTitle
    </p>
    <button
      className="inline ml-auto my-auto"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="stroke-current text-gray-500 dark:text-gray-300 w-5 h-5"
        fill="none"
        height={24}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="3 6 5 6 21 6"
        />
        <path
          d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
        />
        <line
          x1="10"
          x2="10"
          y1="11"
          y2="17"
        />
        <line
          x1="14"
          x2="14"
          y1="11"
          y2="17"
        />
      </svg>
    </button>
  </div>
  <div
    className="grid grid-cols-1 space-y-5"
  >
    <div
      className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
    >
      <div
        className="flex"
      >
        <div
          className="h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
        >
          <svg
            viewBox="0 0 1 1"
          >
            <rect
              fill="#000"
              height={1}
              width={1}
            />
            <text
              dy="0.178"
              fill="#FFF"
              style={
                Object {
                  "fontFamily": "Inter",
                  "fontSize": 0.5,
                  "fontWeight": "bold",
                }
              }
              textAnchor="middle"
              x={0.5}
              y={0.5}
            >
              Pr
            </text>
          </svg>
        </div>
        <div
          className="my-auto inline-block ml-5"
        >
          <p
            className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
          >
            Test title
          </p>
          <p
            className="text-xs font-bold text-gray-500 dark:text-gray-400"
          >
            — primary
          </p>
        </div>
      </div>
      <p
        className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
      >
        <span>
          This is a simple test message.

It spans over multiple lines.
        </span>
      </p>
      <div
        className="flex"
      >
        <div
          className="inline-block"
        >
          <p
            className="text-xs font-bold text-gray-500 dark:text-gray-400"
          >
            — by 
            Primary
          </p>
          <p
            className="text-xs font-bold text-gray-500 dark:text-gray-400"
          >
            2 hours ago
          </p>
        </div>
      </div>
    </div>
    <div
      className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
    >
      <div
        className="flex"
      >
        <div
          className="h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
        >
          <svg
            viewBox="0 0 1 1"
          >
            <rect
              fill="#000"
              height={1}
              width={1}
            />
            <text
              dy="0.178"
              fill="#FFF"
              style={
                Object {
                  "fontFamily": "Inter",
                  "fontSize": 0.5,
                  "fontWeight": "bold",
                }
              }
              textAnchor="middle"
              x={0.5}
              y={0.5}
            >
              Pr
            </text>
          </svg>
        </div>
        <div
          className="my-auto inline-block ml-5"
        >
          <p
            className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
          >
            Test title
          </p>
        </div>
      </div>
      <p
        className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
      >
        <span>
          
        </span>
      </p>
      <div
        className="flex"
      >
        <div
          className="inline-block"
        >
          <p
            className="text-xs font-bold text-gray-500 dark:text-gray-400"
          >
            — by 
            Primary
          </p>
          <p
            className="text-xs font-bold text-gray-500 dark:text-gray-400"
          >
            2 hours ago
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageContainer component renders correctly with no messages 1`] = `
<div
  className="bg-gray-100 dark:bg-gray-700 inline-block p-5 pb-6 rounded-3xl shadow-xl "
>
  <div
    className="flex mb-4 mx-2"
  >
    <p
      className="text-2xl font-bold text-gray-800 dark:text-gray-300"
    >
      TestTitle
    </p>
    <button
      className="inline ml-auto my-auto"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="stroke-current text-gray-500 dark:text-gray-300 w-5 h-5"
        fill="none"
        height={24}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="3 6 5 6 21 6"
        />
        <path
          d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
        />
        <line
          x1="10"
          x2="10"
          y1="11"
          y2="17"
        />
        <line
          x1="14"
          x2="14"
          y1="11"
          y2="17"
        />
      </svg>
    </button>
  </div>
  <div
    className="grid grid-cols-1 space-y-5"
  />
</div>
`;

exports[`MessageContainer component renders correctly with one message 1`] = `
<div
  className="bg-gray-100 dark:bg-gray-700 inline-block p-5 pb-6 rounded-3xl shadow-xl "
>
  <div
    className="flex mb-4 mx-2"
  >
    <p
      className="text-2xl font-bold text-gray-800 dark:text-gray-300"
    >
      TestTitle
    </p>
    <button
      className="inline ml-auto my-auto"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="stroke-current text-gray-500 dark:text-gray-300 w-5 h-5"
        fill="none"
        height={24}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="3 6 5 6 21 6"
        />
        <path
          d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
        />
        <line
          x1="10"
          x2="10"
          y1="11"
          y2="17"
        />
        <line
          x1="14"
          x2="14"
          y1="11"
          y2="17"
        />
      </svg>
    </button>
  </div>
  <div
    className="grid grid-cols-1 space-y-5"
  >
    <div
      className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
    >
      <div
        className="flex"
      >
        <div
          className="h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
        >
          <svg
            viewBox="0 0 1 1"
          >
            <rect
              fill="#000"
              height={1}
              width={1}
            />
            <text
              dy="0.178"
              fill="#FFF"
              style={
                Object {
                  "fontFamily": "Inter",
                  "fontSize": 0.5,
                  "fontWeight": "bold",
                }
              }
              textAnchor="middle"
              x={0.5}
              y={0.5}
            >
              Pr
            </text>
          </svg>
        </div>
        <div
          className="my-auto inline-block ml-5"
        >
          <p
            className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
          >
            Test title
          </p>
          <p
            className="text-xs font-bold text-gray-500 dark:text-gray-400"
          >
            — primary
          </p>
        </div>
      </div>
      <p
        className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
      >
        <span>
          This is a simple test message.

It spans over multiple lines.
        </span>
      </p>
      <div
        className="flex"
      >
        <div
          className="inline-block"
        >
          <p
            className="text-xs font-bold text-gray-500 dark:text-gray-400"
          >
            — by 
            Primary
          </p>
          <p
            className="text-xs font-bold text-gray-500 dark:text-gray-400"
          >
            2 hours ago
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
